providers/implementations/keymgmt/mlx_kmgmt.c

```diff
@@ -0,0 +1,764 @@
+            /* Absent private key data, import public keys */
+            if (!load_slot(key->libctx, key->propq, OSSL_PKEY_PARAM_PUB_KEY,
+                           selection, key, slot, pubenc,
+                           key->minfo->pubkey_bytes, key->xinfo->pubkey_bytes))
+                goto err;
+        }
+    }
+    key->state = prvlen ? MLX_HAVE_PRVKEY : MLX_HAVE_PUBKEY;
+    return 1;
+
+  err:
+    EVP_PKEY_free(key->mkey);
+    EVP_PKEY_free(key->xkey);
+    key->xkey = key->mkey = NULL;
+    key->state = MLX_HAVE_NOKEYS;
+    return 0;
+}
+
+static int mlx_kem_key_fromdata(MLX_KEY *key,
+                               const OSSL_PARAM params[],
+                               int include_private)
+{
+    const OSSL_PARAM *param_prv_key = NULL, *param_pub_key;
+    const void *pubenc = NULL, *prvenc = NULL;
+    size_t pubkey_bytes, prvkey_bytes;
+    size_t publen = 0, prvlen = 0;
+
+    /* Invalid attempt to mutate a key, what is the right error to report? */
+    if (key == NULL || mlx_kem_have_pubkey(key))
+        return 0;
+    pubkey_bytes = key->minfo->pubkey_bytes + key->xinfo->pubkey_bytes;
+    prvkey_bytes = key->minfo->prvkey_bytes + key->xinfo->prvkey_bytes;
+
+    /* What does the caller want to set? */
+    param_pub_key = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PUB_KEY);
+    if (param_pub_key != NULL &&
+        OSSL_PARAM_get_octet_string_ptr(param_pub_key, &pubenc, &publen) != 1)
+        return 0;
+    if (include_private)
+        param_prv_key = OSSL_PARAM_locate_const(params,
+                                                OSSL_PKEY_PARAM_PRIV_KEY);
+    if (param_prv_key != NULL &&
+        OSSL_PARAM_get_octet_string_ptr(param_prv_key, &prvenc, &prvlen) != 1)
+        return 0;
+
+    /* The caller MUST specify at least one of the public or private keys. */
+    if (publen == 0 && prvlen == 0) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_KEY);
+        return 0;
+    }
+
+    /*
+     * When a pubkey is provided, its length MUST be correct, if a private key
+     * is also provided, the public key will be otherwise ignored.  We could
+     * look for a matching encoded block, but unclear this is useful.
+     */
+    if (publen != 0 && publen != pubkey_bytes) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH);
+        return 0;
+    }
+    if (prvlen != 0 && prvlen != prvkey_bytes) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH);
+        return 0;
+    }
+
+    return load_keys(key, pubenc, publen, prvenc, prvlen);
+}
+
+static int mlx_kem_import(void *vkey, int selection, const OSSL_PARAM params[])
+{
+    MLX_KEY *key = vkey;
+    int include_private;
+
+    if (!ossl_prov_is_running() || key == NULL)
+        return 0;
+
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0)
+        return 0;
+
+    include_private = selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY ? 1 : 0;
+    return mlx_kem_key_fromdata(key, params, include_private);
```

Comment:
I'd suggest just passing in the selection to the underlying function rather than just pulling out a single signal?
It it a style issue - but I tend to avoid a single binary flag even in internal functions as things tend to expand over time - and then you have to change later ...

Reply:
This is identical in all the other providers, I'm leaving it be for now.